### PR TITLE
Cleanup and documentation of the interpretation of evaluable references (for unfold)

### LINKED ISSUE #11727: documentation lies about example tactic notation taking a reference
#### Description of the problem

The documentation table under https://coq.inria.fr/refman/user-extensions/syntax-extensions.html#coq:cmd.tactic-notation claims that:

> Tactic argument type | parsed as | interpreted as | as in tactic
> -- | -- | -- | --
> reference | qualified identifier | a global reference of term | unfold

However, `unfold` permits far more than just `reference`.  Here are two examples of things that fail with `reference` but succeed with `unfold`:

```coq
Tactic Notation "myunfold" reference(x) := unfold x.
Notation idnat := (@id nat).
Goal let n := 0 in idnat n = 0.
Proof.
  intro n.
  Fail myunfold idnat.
  (* The command has indeed failed with message:
idnat is bound to a notation that does not denote a reference. *)
  unfold idnat.
  Fail myunfold n.
  (* The command has indeed failed with message:
The reference n was not found in the current environment. *)
  unfold n.
```

cc @herbelin cf https://github.com/coq/coq/pull/11721#issuecomment-593127350 and maybe see also https://github.com/coq/coq/issues/9514

#### Coq Version

8.10.0

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** documentation / cleanup

We clarify the two possible interpretation strategies for evaluable references, such as used in `unfold` or the `delta` flag of `cbv`/`cbn`/`lazy`:
- new type `or_dynamic_binding` to characterize the result of `interning` depending on whether the interpretation is done in an `Ltac` or in a toplevel call to `unfold`;
- relaxed static semantics for this argument: we don't fail statically if it is an axiom or an opaque term, since this status can change at runtime (by functor instantiation or transparent setting, respectively) (see related discussion wrt Ltac2 at #11727);
- new tactic argument `evaluable_ref` precisely matching what `unfold` does, addressing the lying documentation issue #11727.

Fixes / closes #11727

@ejgallego: I put `or_dynamic_binding` in `genarg.mli`. Does it look ok to you?

- [X] Added / updated test-suite
- [X] Corresponding documentation was added / updated
- [X] Entry added in the changelog

### CODE REVIEW

**Comment by ppedrot** [comment]
Gaaah. Are you at least aware that you're making the accidental complexity of the implementation even worse?

**Comment by ppedrot** [comment]
FWIW, my alternative proposal for Ltac2 is the following:
- make the `reference:(...)` quotation dynamically check for variable existence (not sure if worth it?)
- if the user wants to write crap, they can do so by explicitly calling the `VarRef` constructor
- make `unfold` dynamically check for the existence of its argument

No need to introduce mind-boggling contraptions like what this PR does.

**Comment by JasonGross** [comment]
How is `or_dynamic_binding` different from `or_var`?

**Comment by JasonGross** [comment]
@ppedrot I think your proposal is missing a way for user tactics to accept `"+"` to refer to the addition global.  Is your proposal for this extending `reference` to support this?

**Comment by herbelin** [comment]
> How is `or_dynamic_binding` different from `or_var`?

Syntactically they are the same (up to the presence of a location on the ident). Semantically, they are different. The type `or_var` is to distinguish between a meta-variable (i.e. Ltac variable) and a value. The type `or_dynamic_binding` is to tell when a names has to be interpreted dynamically in objects which have hypothesis names as possible values (the dynamic evaluation is only for toplevel commands, not for Ltac functions).

**Comment by ppedrot** [comment]
@JasonGross my proposal is:
- To extend `wit_ref` to also parse "smart globals", i.e. the notation thingy
- Make the static checks on `wit_ref` laxer, by making them accept unknown variables, but checking this at runtime
- Remove the dreaded `Taccoerce.coerce_to_evaluable_ref` and replace it by a check done at the level of tactics

This requires removing the ad-hoc handling of references in `unfold`. This kind of feature, where every hardcoded tactic takes its own little path to decide what is OK and what is not following broken heuristics is an incredible source of non-uniformity and outright bugs.

**Comment by JasonGross** [comment]
> Make the static checks on wit_ref laxer, by making them accept unknown variables, but checking this at runtime

Is this "at runtime uniformly in all users of wit_ref", or "at runtime in each tactic separately"?  (I would prefer the latter; I don't want it to be a requirement that I can only is wit_ref for things that can be evaluated)

**Comment by ppedrot** [comment]
Checking for variable existence should be done uniformly at runtime, but "evaluability" needs to be done tactic-wise.

File: doc/sphinx/user-extensions/syntax-extensions.rst [hunk 1]
```diff
@@ -1834,6 +1839,18 @@ Tactic notations allow to customize the syntax of tactics. They have the followi
              underlying entry can be used: entry is either ``constr``, ``hyp``, ``integer``
              or ``int_or_var``.
 
+   .. note:: A *smart* qualified identifier is:
+
+             - either a qualified identifier (possibly with empty qualification)
```
Comment (jfehrle) [inline doc/sphinx/user-extensions/syntax-extensions.rst]: Let's avoid defining syntax in text where possible.  What part of the grammar does this correspond to?  Let's hyperlink to the syntax or show it nearby.

**Comment by herbelin** [comment]
@ppedrot:

>  Make the static checks on `wit_ref` laxer, by making them accept unknown variables, but checking this at runtime

What do you mean by unknown variables? Do you mean the permissivity of the non strict check mode (i.e. toplevel mode) to refer to hypotheses which are not statically bound to an ltacX variable but only present in the typing context at the time of execution?

>  Remove the dreaded `Taccoerce.coerce_to_evaluable_ref` and replace it by a check done at the level of tactics
> 
> This requires removing the ad-hoc handling of references in `unfold`. This kind of feature, where every hardcoded tactic takes its own little path to decide what is OK and what is not following broken heuristics is an incredible source of non-uniformity and outright bugs.

What do you mean by "a check done at the level of tactics"? Do you mean at the level of an `Ltac2` tactic definition? Or at the level of the tactics in `tactics.ml`?

What would change compared to Ltac1? For instance, how would you translate the subtyping in, say, `Ltac f x c := set (x := c); unfold x`?

File: doc/sphinx/user-extensions/syntax-extensions.rst (hunk 1 above)
Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: I suspect that the grammar is not explicitly highlighted anywhere. It appears for instance at the end of the documentation for [`unfold`](https://coq.inria.fr/refman/proof-engine/tactics.html?highlight=unfold#coq:tacn.unfold).

Comment (jfehrle) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: OK, maybe just add a todo in the .rst with the reference, we'll fix it later.

Comment (Zimmi48) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: @jfehrle It is `smart_qualid` (used to be `smart_global` which we recently renamed).

**Comment by ppedrot** [comment]
> Do you mean the permissivity of the non strict check mode

Yes.

>  Or at the level of the tactics in tactics.ml?

That one, so as to share the checks.

**Comment by herbelin** [comment]
> > Do you mean the permissivity of the non strict check mode
> 
> Yes.
> 
> > Or at the level of the tactics in tactics.ml?
> 
> That one, so as to share the checks.

It is then not crystal clear to me what you have in mind and how very different from the current situation it would be. In particular:

> make the static checks on `wit_ref` laxer, by making them accept unknown variables, but checking this at runtime

As far as I know, this is what is already done (even if in a convoluted way).

> Remove the dreaded Taccoerce.coerce_to_evaluable_ref and replace it by a check done at the level of tactics

Is this a change of semantics or just an architectural change?

----

But anyway, if we are ok to:
- not add `evaluable_ref` (and then remove the check statically failing on `IndRef` and `ConstructRef` in Ltac-definition-mode for `unfold`; since, after all, this is the only difference between `smart_global` and the particular case of `smart_global` for evaluable globals);
- instead extend `reference` to interning `smart_global` and document it.

Then, I'm going to do it.

File: doc/sphinx/user-extensions/syntax-extensions.rst (hunk 1 above)
Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: OK, thanks. Then will use it.

**Comment by ppedrot** [comment]
@herbelin your plan seems OK to me.

**Comment by herbelin** [comment]
> But anyway, if we are ok to:
> - not add `evaluable_ref` (and then remove the check statically failing on `IndRef` and `ConstructRef` in Ltac-definition-mode for `unfold`; since, after all, this is the only difference between `smart_global` and the particular case of `smart_global` for evaluable globals);
> - instead extend `reference` to interning `smart_global` and document it.

Some more comments on this plan:
- There are several other use of `global`or `reference` in commands. For some of them, it may be relevant to want a smart global:
  - `Extraction global` and its variants: are we ready to accept to write `Extraction "+"`?
  - `Numeral Notation global global global`/`String Notation global global global`: are we ready to accept things like `Numeral Notation "ℕ" "⟦ _ ⟧" "〈 _ 〉"`?
  - `firstorder using reference`: why not accepting `firstorder using "⚬"` for instance (note in passing that `auto using` uses constr instead; do we uniformize?)
  - `Show Match reference`: do we accept `Show Match "ℕ"`? Why not.

  But for other ones, it does not make sense to want a smart global:
  - `Print Ltac reference`/`Print Ltac2 reference`: we don't have notations for LtacX references, so, I believe one should keep a variant of `global`/`reference` which is not smart (namely `qualid`?)
  - `Print Section global`/`Print Module global`, `Print Library global`: a smart global is not relevant either.

- As a conclusion, I believe we should keep a difference between those `reference` which denote a reference to a term and other usages of `reference` for tactics, modules, etc. There are two possibilities:
  - either we use a new name for non-constr reference, say `pure_reference`, or `qualid`, ...
  - or we rename this extended `reference` argument for `Tactic Notation`/`Ltac2 Notation`/`TACTIC EXTEND`/`VERNAC EXTEND` into say `smart_reference`?

  I'm currenlty leaning for option 1.

Otherwise, I must say after a few days of resting that:
- I'm not fully convinced that we should minimize the number of generically-managed types of LtacX arguments vs manage at the level of a given tactic what exact subset of a generic argument it accepts. I'm ok with having found a compromise (which I'm implementing) but I'm not convinced that this is the best thing to do.
- I'm a bit afraid to allow more uses of the syntax `"+"`, considering that it has a drawback, even if it has the clear advantage to directly echo to the syntax of `Notation`. The drawback is:
  - that it may be confusing in places where a string can also be given (as it is already the case for  `simpl` which accepts both `simpl pattern` and `simpl reference`);
  - and in particular, that it is not usable within a term.

  I don't have a good alternative though:
  - Agda's use of `_` would be quite different in style to the format of notations we are already used to (moreover how to escape `_` when it appears in the notation?).
  - OCaml's use of `( )` does not really scale to misfix notations.

  On the other side, we can imagine qualifying it, as in `Top."+"`. So, maybe should I just be happy with the double quotes. And, maybe could we just solve the use of global-as-notation in terms by requiring writing `("{ _ } + { _ }")` in terms, like in OCaml.

Any opinions?

TL;DR

In the absence of comments, I'm leaning towards continuing exploring the above direction:
- keeping `reference` to denote something which is parsed as a smart qualid and interpreted as a global reference, adding in passing support for smart references in `Extraction`, `Numeral`, `String`;
- use `qualid` for current non-term uses of `reference`.

**Comment by Zimmi48** [comment]
It seems to me that this PR should be closed and turned into an issue since it is now a plan that is getting implemented in other PRs. Just removing the milestone for now.

**Comment by herbelin** [comment]
I'll have to check, but some parts of this is not #12254 and #12256.

**Comment by JasonGross** [comment]
@herbelin I've run into some confusion at https://github.com/coq/coq/pull/12512#issuecomment-644999726.  Some tactics interpret aliases as referring to the constant at their head, while some do not.  Consider:
```coq
Notation p a b := (Nat.add a b) (only parsing).
Print p. (* Notation p a b := (a + b) *)
Goal p 0 0 = p 0 0.
  cbv delta [p]. (* success *)
```
Does this difference between `Print` and `cbv delta` have anything to do with the cleanup here?  Or is it something orthogonal?

**Comment by Blaisorblade** [comment]
My comments at https://github.com/coq/coq/pull/12512#issuecomment-645007190 might be relevant, but I'm too hopelessly lost to turn all that confusion into something intelligent...

**Comment by maximedenes** [comment]
What's the status of this PR? Should it be reviewed?

**Comment by herbelin** [comment]
@jasongross:
> Does this difference between `Print` and `cbv delta` have anything to do with the cleanup here? Or is it something orthogonal?

As discussed in the call, this is rather orthogonal.

@Blaisorblade 
> My comments at https://github.com/coq/coq/pull/12512#issuecomment-645007190 might be relevant

Yes, this is related.

@maximedenes
> What's the status of this PR? Should it be reviewed?

I just took the time to come to it. Leaving away a few experiments, what only remains interesting is a few more `unfold` tests and fixing the manual which still says that `unfold` is an example of tactics taking a `qualid` as arguments.

I'm closing it, opening instead the (very light) #12698.
